Add BlockHash for 3 BytesRefs

dnhatn opened the pull request:
This change introduces a specialized BlockHash for 3 BytesRefs. Unlike other specialized BlockHashes, this BlockHash should handle nulls and multivalues properly. While not intended as a showcase, it can illustrate the performance enhancements of ESQL on multi-term aggregations over the search API in our benchmark. This change is expected to reduce the execution time of multi-term aggregations from 2054ms to 1178ms.

dnhatn committed:
Add BlockHash for 3 BytesRefs

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BytesRef3BlockHash.java:
```diff
@@ -0,0 +1,220 @@
+
+    @Override
+    public ReleasableIterator<IntBlock> lookup(Page page, ByteSizeValue targetBlockSize) {
+        throw new UnsupportedOperationException("TODO");
```
Sorry @nik9000. I will implement this in a follow-up.

nik9000 replied:
None of the specialize versions have an implementation of it at the moment. It's kind of a pain to implement. I've got to make some building blocks to help it.

elasticsearchmachine commented:
Hi @dnhatn, I've created a changelog YAML for you.

dnhatn committed:
Update docs/changelog/108165.yaml

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

nik9000 commented:
> This change is expected to reduce the execution time of multi-term aggregations from 2054ms to 1178ms.

Nice.

I wonder how much we're leaving on the floor because `PackedValuesBlockHash` is just not all that efficient. It works, but we never tried to make it as fast as possible. And I guess that's ok because it has to be the catch all. But I'm sure it could do better.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BytesRef3BlockHash.java:
```diff
@@ -0,0 +1,220 @@
+    private void addVectors(BytesRefVector v1, BytesRefVector v2, BytesRefVector v3, GroupingAggregatorFunction.AddInput addInput) {
+        final int positionCount = v1.getPositionCount();
+        try (IntVector.Builder ordsBuilder = blockFactory.newIntVectorFixedBuilder(positionCount)) {
+            // TODO: enable ordinal vectors in BytesRefBlockHash
```
One day I'd like to do away with the special ordinals aggregator and use ordinals here. But that's a problem for another day.

dnhatn replied:
+1

nik9000 commented on x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/aggregation/blockhash/BytesRefBlockHash.java:
```diff
@@ -87,6 +88,9 @@ IntVector add(BytesRefVector vector) {
     }
 
     IntBlock add(BytesRefBlock block) {
+        if (block instanceof OrdinalBytesRefBlock ordinalsBlock) {
```
Could you make an `asOrdinals` instead? I like the `asVector` pattern we have now. It's nice looking at `Block` and seeing "oh, you can view this as a Vector sometimes". And this'd show the same thing.

dnhatn replied:
Yes, I added `asOrdinals` in https://github.com/elastic/elasticsearch/pull/108165/commits/a455c55a9d73f38e155afbbb13cd938d31d69a1b.

nik9000 replied:
:heart:

nik9000 commented on x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/blockhash/BlockHashTests.java:
```diff
@@ -1202,7 +1296,9 @@ public void add(int positionOffset, IntVector groupIds) {
-        if (blockHash instanceof LongLongBlockHash == false && blockHash instanceof BytesRefLongBlockHash == false) {
+        if (blockHash instanceof LongLongBlockHash == false
+            && blockHash instanceof BytesRefLongBlockHash == false
+            && blockHash instanceof BytesRef3BlockHash == false) {
```
I think this shouldn't be in the `instanceof` because it does do the right thing with nulls.

dnhatn replied:
This is to skip the assert lookup.

nik9000 commented on server/src/main/java/org/elasticsearch/common/util/Int3Hash.java:
```diff
@@ -0,0 +1,141 @@
+    private long set(int key1, int key2, int key3, long id) {
+        assert size < maxSize;
+        final long slot = slot(hash(key1, key2, key3), mask);
+        for (long index = slot;; index = nextSlot(index, mask)) {
```
You know I didn't write it because it has stuff like `for(;;)`. It's one of the few stylistic fingerprints we have. I'll always use a `while` loop, sometimes even `while(true)` to put you on notice "something weird is happening here". The two `;`s next to each other does the same. And it's much more your style. It's nice to have some creative freedom.

dnhatn replied:
I copied these from LongLongHash. I switched to while style in https://github.com/elastic/elasticsearch/pull/108165/commits/ca42f3b2466066b433b1144e3617eba584257666.

nik9000 replied:
I wonder who's style the funny `for` is? It's fine either way.

dnhatn replied:
I think that was Uri :) (per git blame).

nik9000 commented:
This also adds an ordinals optimization to `BlockHash`. I'm sure it's nice as well.

dnhatn committed:
asOrdinals()

dnhatn committed:
style

dnhatn committed:
skip lookup

dnhatn committed:
doc

dnhatn commented:
@nik9000 Thank you for your review. I think we have addressed your comments.

> I wonder how much we're leaving on the floor because PackedValuesBlockHash is just not all that efficient. It works, but we never tried to make it as fast as possible. And I guess that's ok because it has to be the catch all. But I'm sure it could do better.

I agree. I think we have some ideas to make PackedValuesBlockHash more efficient, but we haven't tried them yet. I will spend some time on these after the tsdb integration.

dnhatn committed:
Use asOrdinals() in enrich

nik9000 reviewed (approved):
thanks for the improvements! :rocket:

dnhatn commented:
Thanks Nik!
